File: app/schemas/reports.py
from datetime import date, datetime
from pydantic import BaseModel, Field, ConfigDict


class ReportEnvelope(BaseModel):
    data: list
    total: int
    page: int
    page_size: int
    generated_at: datetime


class DateRangeFilter(BaseModel):
    preset: str | None = None
    from_date: date | None = None
    to_date: date | None = None
    shift: str | None = None
    group_by: str | None = None


class ReportPagination(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=50, ge=1, le=500)


class RouteDeliveryItem(BaseModel):
    route_id: int
    route_name: str
    route_code: str
    session_count: int
    delivery_count: int
    total_loaded_quantity: float
    total_delivered_quantity: float
    total_cash_collected: float
    total_token_registered: float
    total_returned_quantity: float
    shortage_surplus: float
    is_balanced: bool
    model_config = ConfigDict(from_attributes=True)


class RouteDeliveryReport(BaseModel):
    route_id: int | None
    date_from: date
    date_to: date
    shift: str | None
    items: list[RouteDeliveryItem]
    summary: RouteDeliveryItem


class RevenueBreakdown(BaseModel):
    source: str
    payment_mode: str | None = None
    route_name: str | None = None
    milk_type_name: str | None = None
    amount: float
    percentage: float
    model_config = ConfigDict(from_attributes=True)


class RevenueReport(BaseModel):
    date_from: date
    date_to: date
    total_revenue: float
    token_book_revenue: float
    customer_bill_revenue: float
    by_source: list[RevenueBreakdown]
    by_payment_mode: list[RevenueBreakdown]
    by_route: list[RevenueBreakdown]
    by_milk_type: list[RevenueBreakdown]


class CustomerCollectionItem(BaseModel):
    customer_id: int
    customer_code: str
    customer_name: str
    route_name: str
    total_billed: float
    total_paid: float
    balance: float
    collection_percentage: float
    last_bill_date: date | None = None
    last_payment_date: date | None = None
    aging_current: float = 0
    aging_31_60: float = 0
    aging_61_90: float = 0
    aging_90_plus: float = 0
    model_config = ConfigDict(from_attributes=True)


class CollectionEfficiencyReport(BaseModel):
    date_from: date
    date_to: date
    route_id: int | None
    total_billed: float
    total_paid: float
    total_balance: float
    overall_collection_percentage: float
    items: list[CustomerCollectionItem]


class ConsumptionDay(BaseModel):
    date: date
    total_quantity: float
    by_milk_type: list[dict]


class ConsumptionTrend(BaseModel):
    period: str
    recent_7day_avg: float | None = None
    preceding_21day_avg: float | None = None
    change_percentage: float | None = None


class CustomerConsumptionReport(BaseModel):
    customer_id: int
    customer_name: str
    date_from: date
    date_to: date
    group_by: str
    total_consumption: float
    average_daily: float
    days_with_data: int
    trend: ConsumptionTrend
    items: list[ConsumptionDay]


class TokenUtilizationItem(BaseModel):
    customer_id: int
    customer_name: str
    route_name: str
    token_number: int
    milk_type_name: str
    total_books_issued: int
    active_books: int
    completed_books: int
    total_sheets_used: int
    total_sheets_remaining: int
    utilization_percentage: float
    books_below_20_percent: int
    model_config = ConfigDict(from_attributes=True)


class TokenUtilizationReport(BaseModel):
    route_id: int | None
    total_customers_with_tokens: int
    total_books_issued: int
    total_sheets_used: int
    total_sheets_remaining: int
    overall_utilization_percentage: float
    items: list[TokenUtilizationItem]


class OperationalDashboard(BaseModel):
    report_date: date
    total_sessions: int
    total_milk_loaded: float
    total_milk_delivered: float
    total_cash_collected: float
    deliveries_by_status: dict
    pending_token_count: int
    unclosed_sessions: int
    unbalanced_sessions: int
    completed_not_closed: int

File: frontend/src/types/reports.ts
export type ReportPreset =
  | "today"
  | "yesterday"
  | "this_week"
  | "last_week"
  | "this_month"
  | "last_month"
  | "this_year";

export interface ReportEnvelope<T> {
  data: T[];
  total: number;
  page: number;
  page_size: number;
  generated_at: string;
}

export interface ReportDateParams {
  preset?: ReportPreset;
  from_date?: string;
  to_date?: string;
  refresh?: boolean;
}

export interface RouteDeliveryParams extends ReportDateParams {
  route_id?: number;
  shift?: string;
}

export interface RevenueParams extends ReportDateParams {
  route_id?: number;
  milk_type_id?: number;
  payment_mode?: string;
  group_by?: string;
}

export interface ConsumptionParams extends ReportDateParams {
  group_by?: string;
}

export interface TokenUtilizationParams {
  route_id?: number;
  customer_id?: number;
  low_threshold?: number;
  refresh?: boolean;
}

export interface CollectionEfficiencyParams extends ReportDateParams {
  route_id?: number;
  min_outstanding?: number;
}

export type DeliveryStatusKey =
  | "DELIVERED"
  | "PENDING_TOKEN"
  | "CASH_SALE"
  | "NOT_DELIVERED"
  | "CANCELLED";

export interface OperationalDashboard {
  report_date: string;
  total_sessions: number;
  total_milk_loaded: number;
  total_milk_delivered: number;
  total_cash_collected: number;
  deliveries_by_status: Record<DeliveryStatusKey, number>;
  pending_token_count: number;
  unclosed_sessions: number;
  unbalanced_sessions: number;
  completed_not_closed: number;
}

export interface RouteDeliveryItem {
  route_id: number;
  route_name: string;
  route_code: string;
  session_count: number;
  delivery_count: number;
  total_loaded_quantity: number;
  total_delivered_quantity: number;
  total_cash_collected: number;
  total_token_registered: number;
  total_returned_quantity: number;
  shortage_surplus: number;
  is_balanced: boolean;
}

export interface RevenueBreakdown {
  source: string;
  payment_mode: string | null;
  route_name: string | null;
  milk_type_name: string | null;
  amount: number;
  percentage: number;
}

export interface RevenueReport {
  date_from: string;
  date_to: string;
  total_revenue: number;
  token_book_revenue: number;
  customer_bill_revenue: number;
  by_source: RevenueBreakdown[];
  by_payment_mode: RevenueBreakdown[];
  by_route: RevenueBreakdown[];
  by_milk_type: RevenueBreakdown[];
}

export interface ConsumptionDay {
  date: string;
  total_quantity: number;
  by_milk_type: { milk_type: string; quantity: number }[];
}

export interface ConsumptionTrend {
  period: string;
  recent_7day_avg: number | null;
  preceding_21day_avg: number | null;
  change_percentage: number | null;
}

export interface CustomerConsumptionReport {
  customer_id: number;
  customer_name: string;
  date_from: string;
  date_to: string;
  group_by: string;
  total_consumption: number;
  average_daily: number;
  days_with_data: number;
  trend: ConsumptionTrend;
  items: ConsumptionDay[];
}

export interface TokenUtilizationItem {
  customer_id: number;
  customer_name: string;
  route_name: string;
  token_number: number;
  milk_type_name: string;
  total_books_issued: number;
  active_books: number;
  completed_books: number;
  total_sheets_used: number;
  total_sheets_remaining: number;
  utilization_percentage: number;
  books_below_20_percent: number;
}

export interface CustomerCollectionItem {
  customer_id: number;
  customer_code: string;
  customer_name: string;
  route_name: string;
  total_billed: number;
  total_paid: number;
  balance: number;
  collection_percentage: number;
  last_bill_date: string | null;
  last_payment_date: string | null;
  aging_current: number;
  aging_31_60: number;
  aging_61_90: number;
  aging_90_plus: number;
}
